src/Common/Core/Impl/Security/SecurityUtilities.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Security;
+
+namespace Microsoft.Common.Core.Security {
+    public static class SecurityUtilities {
+        public static IntPtr CreatePasswordBuffer() {
+            return CreateSecureStringBuffer(NativeMethods.CREDUI_MAX_PASSWORD_LENGTH);
+        }
+
+        public static SecureString ToSecureString(this string s) {
```

Comment:
Yes, this is the root of the problem. It's not even about `Find`, fundamentally. The problem is that when HTTP request comes in and is parsed by Kestrel, the result of that parse is a bunch of non-secure strings containing all the header data - and that will include the credentials. So it doesn't really matter how far we push `SecureString` in our code, at some point we're going to run into that boundary, and we don't control the code on the other side of it.

We will likely need a custom authentication method that uses some sort of digest based auth (like HTTP Digest, but using something more secure than MD5), so that the password is secure as it flows through the request pipeline not under our control. Then when we decrypt it, it goes into `SecureString` immediately. The nonce should also be cleared immediately as soon as password is decrypted.

I'm not sure how to make this play well with `ClaimsCredential`, though. I think all claims are plain strings.